V1.0.0 RC (#182)
OWASP/maswe: weaknesses/MASVS-PLATFORM/MASWE-0029.md

Reviewer: Jacobo: DONE

---
title: Unnecessary Exposure of Sensitive Data via Notifications
id: MASWE-0029
alias: data-leak-notifications
requirement: "The app does not unnecessarily expose sensitive data through system notifications."
platform: [android, ios]
profiles: [L2]
threat: MAS-THREAT-0029
attacks: [MAS-ATTACK-0044, MAS-ATTACK-0045]
mappings:
  masvs-v2: [MASVS-PLATFORM-3, MASVS-STORAGE-2]
  cwe: [200, 359]
  maswe-beta: [MASWE-0054]
refs:
- https://developer.android.com/develop/ui/views/notifications/build-notification#lockscreenNotification
- https://developer.apple.com/documentation/usernotifications
status: new
---

## Overview

This weakness occurs when an app includes more sensitive data (such as one-time codes, message contents, or account details) than necessary in a system notification.

> Notifications can be rendered on the lock screen, and therefore, anyone holding the device can read them without unlocking it. They also can be read by other apps that hold notification access (e.g. an Android [`NotificationListenerService`](https://developer.android.com/reference/android/service/notification/NotificationListenerService)). Any sensitive value included in a notification therefore leaves the app's control the moment it is posted.

Reviewer: ```suggestion
Notifications can be rendered on the lock screen, and therefore, anyone holding the device can read them without unlocking it. They also can be read by other apps that hold notification access, if the platform allows it (e.g. an Android [`NotificationListenerService`](https://developer.android.com/reference/android/service/notification/NotificationListenerService)). Any sensitive value included in a notification therefore leaves the app's control the moment it is posted.
```

## Modes of Introduction

- **Sensitive Content in Notifications**: Including one-time codes, message contents, financial details, or other sensitive values directly in notification titles or bodies when a generic notification would suffice.

> - **No Lock-Screen Redaction**: Not configuring notification visibility so that sensitive content is redacted or hidden on the lock screen.

Reviewer: @cpholguera we can add a mode of introduction here (for the moment) which can be something like "Unappropiate management of user input in the notification content", pointing to https://github.com/OWASP/masvs/discussions/772. The thing is that the discussion is not talking about sensitive data leaked in notifications, but the possibility of hiding the data from the notification preview by using a long user input. Maybe adding this in the MASWE related to handling user input is better?

Author: Probably yes, thanks!

Author: That'd be https://github.com/OWASP/maswe/pull/182/changes#diff-5663fe24070bf6cb402d8e0d037d9bac391996e0cdffac8f153a17c1fffc6d9b could you please propose something in there?

## Impact

- **Compromise of Sensitive Data**: Unauthorized observers or, where applicable, apps with notification-listener access may obtain personal, financial, authentication, or account information contained in notifications.
- **Authentication or Authorization Bypass**: Attackers may use exposed credentials or still-valid one-time codes to access an account, perform unauthorized transactions, or carry out other sensitive actions.

## Mitigations

> - **Minimize Notification Content**: Use notifications only to indicate that an event occurred, and retrieve sensitive details inside the app after appropriate authentication and authorization. For iOS remote notifications, follow Apple's [remote-notification payload guidance](https://developer.apple.com/documentation/usernotifications/generating-a-remote-notification) and keep user-visible alert content non-sensitive by default.

Reviewer: ```suggestion
- **Minimize Notification Content**: Use notifications only to indicate that an event occurred, and retrieve sensitive details inside the app after appropriate authentication and authorization. Do not place sensitive data in the title because private notifications may still display basic information including the title.
```

> - **Configure Android Lock-Screen Visibility**: Use [`NotificationCompat.Builder.setVisibility()` and `setPublicVersion()`](https://developer.android.com/develop/ui/views/notifications/build-notification#lockscreenNotification) to modify the presentation of sensitive notifications. Use `VISIBILITY_PRIVATE` together with a generic public version when a redacted notification may be shown, and use `VISIBILITY_SECRET` when the notification should not appear on a secure lock screen. Do not place sensitive data in the title because private notifications may still display basic information including the title.

Reviewer: ```suggestion
```

Reviewer: Removed and used a more generic recommendation

Author: Even if it may be an Android-only mitigation, isn't "Configure Lock-Screen Visibility" is still valuable as a separate mitigation separate from "Minimize Notification Content" (which would apply to both OSs)?

> - **Configure Android Lock-Screen Visibility**: Use [`NotificationCompat.Builder.setVisibility()` and `setPublicVersion()`](https://developer.android.com/develop/ui/views/notifications/build-notification#lockscreenNotification) to modify the presentation of sensitive notifications. Use `VISIBILITY_PRIVATE` together with a generic public version when a redacted notification may be shown, and use `VISIBILITY_SECRET` when the notification should not appear on a secure lock screen. Do not place sensitive data in the title because private notifications may still display basic information including the title.

Author: ```suggestion
- **Configure Lock-Screen Visibility**: Use [`NotificationCompat.Builder.setVisibility()` and `setPublicVersion()`](https://developer.android.com/develop/ui/views/notifications/build-notification#lockscreenNotification) to modify the presentation of sensitive notifications. Use `VISIBILITY_PRIVATE` together with a generic public version when a redacted notification may be shown, and use `VISIBILITY_SECRET` when the notification should not appear on a secure lock screen. Do not place sensitive data in the title because private notifications may still display basic information including the title.
```

---

> title: Sensitive Data Leaked via Notifications

Reviewer: ```suggestion
title: Unnecessary Exposure of Sensitive Data via Notifications
```

id: MASWE-0029
alias: data-leak-notifications
requirement: "The app does not unnecessarily expose sensitive data through system notifications."
platform: [android, ios]
profiles: [L2]
threat: MAS-THREAT-0029
attacks: [MAS-ATTACK-0044, MAS-ATTACK-0045]
mappings:
  masvs-v2: [MASVS-PLATFORM-3, MASVS-STORAGE-2]
  cwe: [200, 359]
  maswe-beta: [MASWE-0054]
refs:
- https://developer.android.com/develop/ui/views/notifications/build-notification#lockscreenNotification
- https://developer.apple.com/documentation/usernotifications
status: new
---

## Overview

> This weakness occurs when an app places sensitive data, such as one-time codes, message contents, or account details, into system notifications without restricting where and to whom they are visible.

Reviewer: ```suggestion
This weakness occurs when an app includes more sensitive data (such as one-time codes, message contents, or account details) than necessary in a system notification.
```

> Notifications are rendered on the lock screen, where anyone holding the device can read them without unlocking it, and they can be read by other apps that hold notification access (e.g. an Android `NotificationListenerService`). Any sensitive value included in a notification therefore leaves the app's control the moment it is posted.

Reviewer: ```suggestion
Notifications can be rendered on the lock screen, and therefore, anyone holding the device can read them without unlocking it. They also can be read by other apps that hold notification access (e.g. an Android [`NotificationListenerService`](https://developer.android.com/reference/android/service/notification/NotificationListenerService)). Any sensitive value included in a notification therefore leaves the app's control the moment it is posted.
```

## Modes of Introduction

> - **Sensitive Content in Notifications**: Including one-time codes, message contents, financial details, or other sensitive values directly in notification titles or bodies.

Reviewer: ```suggestion
- **Sensitive Content in Notifications**: Including one-time codes, message contents, financial details, or other sensitive values directly in notification titles or bodies when a generic notification would suffice.
```

- **No Lock-Screen Redaction**: Not configuring notification visibility so that sensitive content is redacted or hidden on the lock screen.

## Impact

> - **Compromise of Sensitive Data**: Attackers can read message contents or account details from notifications, resulting in unauthorized disclosure of user data.

Reviewer: ```suggestion
- **Compromise of Sensitive Data**: Unauthorized observers or, where applicable, apps with notification-listener access may obtain personal, financial, authentication, or account information contained in notifications.
```

> - **Authentication or Authorization Bypass**: Attackers can capture one-time codes delivered via notifications, resulting in the defeat of SMS- or push-based authentication factors and unauthorized account access.

Reviewer: ```suggestion
- **Authentication or Authorization Bypass**: Attackers may use exposed credentials or still-valid one-time codes to access an account, perform unauthorized transactions, or carry out other sensitive actions.
```

## Mitigations

> - **Keep Sensitive Data Out of Notifications**: Use notifications to signal that something happened and reveal the sensitive details only inside the app after unlocking it.

Reviewer: ```suggestion
- **Minimize Notification Content**: Use notifications only to indicate that an event occurred, and retrieve sensitive details inside the app after appropriate authentication and authorization. For iOS remote notifications, follow Apple's [remote-notification payload guidance](https://developer.apple.com/documentation/usernotifications/generating-a-remote-notification) and keep user-visible alert content non-sensitive by default.
```

> - **Redact Notifications on the Lock Screen**: Configure notification visibility (e.g. private visibility with a public redacted version on Android) so sensitive content is hidden until the device is unlocked.

Reviewer: ```suggestion
- **Configure Android Lock-Screen Visibility**: Use [`NotificationCompat.Builder.setVisibility()` and `setPublicVersion()`](https://developer.android.com/develop/ui/views/notifications/build-notification#lockscreenNotification) to modify the presentation of sensitive notifications. Use `VISIBILITY_PRIVATE` together with a generic public version when a redacted notification may be shown, and use `VISIBILITY_SECRET` when the notification should not appear on a secure lock screen. Do not place sensitive data in the title because private notifications may still display basic information including the title.
```

> - **Redact Notifications on the Lock Screen**: Configure notification visibility (e.g. private visibility with a public redacted version on Android) so sensitive content is hidden until the device is unlocked.

Reviewer: ```suggestion
- **Configure iOS Hidden-Preview Content Safely**: For categories that may contain sensitive information, provide a generic [`hiddenPreviewsBodyPlaceholder`](https://developer.apple.com/documentation/usernotifications/unnotificationcategory/hiddenpreviewsbodyplaceholder). Do not enable [`hiddenPreviewsShowTitle`](https://developer.apple.com/documentation/usernotifications/unnotificationcategoryoptions/hiddenpreviewsshowtitle) or [`hiddenPreviewsShowSubtitle`](https://developer.apple.com/documentation/usernotifications/unnotificationcategoryoptions/hiddenpreviewsshowsubtitle) unless those fields are safe to display. These category options customize the presentation used when previews are already hidden; they do not force previews to be hidden. The app may direct users to its notification settings using [`openNotificationSettingsURLString`](https://developer.apple.com/documentation/uikit/uiapplication/opennotificationsettingsurlstring), but user guidance is not a substitute for safe notification content.
```

Reviewer: added specific reccomendations for each platform.

> - **Review Notification Content Regularly**: Audit which notifications the app posts and ensure new features do not introduce sensitive values into them.

Reviewer: ```suggestion
```

Reviewer: I wouldn't put a contiunous review suggestion, just applying the previous one would suffice.
